### PR TITLE
Add never type to `TypeName`

### PR DESCRIPTION
**Pull request by nathanwhit**
Addresses part of #368

### CODE REVIEW

**Comment by nathanwhit** [comment]
> Copy/clone are in libcore: https://github.com/rust-lang/rust/blob/master/src/librustc_trait_selection/traits/select.rs#L2255
> 
> Also, looking over the table, looks like raw ptrs and immutable refs also have their impls in libcore. Think you can fix that here too? (It's small enough I don't think we _need_ to split into a separate PR)

Whoops you're right, should be fixed now!

File: chalk-solve/src/clauses/builtin_traits/sized.rs [hunk 1]
```diff
@@ -74,9 +74,7 @@ pub fn add_sized_program_clauses<I: Interner>(
             TypeName::Tuple(arity) => {
                 push_tuple_sized_conditions(db, builder, trait_ref, *arity, substitution)
             }
-            TypeName::Scalar(_) | TypeName::Raw(_) | TypeName::Ref(_) => {
```
Comment (jackh726) [inline chalk-solve/src/clauses/builtin_traits/sized.rs:77]: These should stay. Only copy/clone is provided.

File: book/src/clauses/well_known_traits.md [hunk 2]
```diff
@@ -34,6 +34,7 @@ Some common examples of auto traits are `Send` and `Sync`.
 | structs         | ⚬    | ⚬    |  ✅    | ✅     | ⚬    | ⚬  |  ⚬      |  ⚬       |   ✅         |
 | scalar types    | 📚    | 📚    | ✅     | ⚬     | ⚬   |  ⚬  |  ⚬     |  ⚬       |    ❌        |
 | str             | 📚    | 📚    | ⚬     | ⚬     | ⚬   |  ⚬  |  ⚬     |  ⚬       |    ❌        |
+| never type      | ✅   |  ✅   |  ✅   |  ⚬    |  ⚬    | ⚬   |   ⚬    |  ⚬       |   ❌       |
 | trait objects   | ⚬    | ⚬    | ⚬     |  ✅    | ⚬    | ⚬   | ⚬      |  ⚬       |    ⚬        |
 | functions ptrs  | ✅    | ✅    | ✅     | ⚬     | ⚬    | ❌   | ⚬      |  ⚬       |    ❌         |
 | raw ptrs        | ✅   |  ✅  |   ✅   |  ⚬    |   ⚬   |  ⚬  |   ⚬    |   ⚬      |      ❌      |
```
Comment (jackh726) [inline book/src/clauses/well_known_traits.md:40]: Can you update this table for copy/clone of raw ptrs and immutable refs?

File: tests/test/refs.rs [hunk 3]
```diff
@@ -51,71 +51,7 @@ fn mut_refs_are_sized() {
...
-fn mut_refs_are_not_clone() {
-    test! {
-        program {
-            #[lang(clone)]
-            trait Clone { }
-        }
-
-        goal {
-            forall<'a, T> { not { &'a mut T: Clone } }
+            forall<'a, T> { &'a T: Sized }
```
Comment (jackh726) [inline tests/test/refs.rs:54]: This should be `&'a mut T`

**Pull request merged**